working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1528,41 +1564,65 @@ The context type schema for a pattern `p` is:
         //                                 ^- Infers List<int>.
         ```
 
-*   **Relational** or **cast**: The context type schema is `Object?`.
+*   **Relational** or **cast**: The context type schema is `?`.
 
 *   **Parenthesized**: The context type schema of the inner subpattern.
 
 *   **List**: A context type schema `List<E>` where:
 
     1.  If `p` has a type argument, then `E` is the type argument.
 
-    2.  Else if `p` has no elements then `E` is `Object?`. *If the pattern
-        doesn't destructure anything, it matches any list, so it is permissive
-        with the context type.*
+    2.  Else if `p` has no elements then `E` is `?`.
 
-    3.  Else `E` is the greatest lower bound of the type schemas of all element
-        subpatterns. *We use the greatest lower bound to ensure that the outer
-        collection type has a precise enough type to ensure that any typed field
-        subpatterns do not need to downcast:*
+    3.  Else, infer the type schema from the subpatterns:
 
-        ```dart
-        var [int a, num b] = [1, 2];
-        ```
+        1.  Let `es` be an empty list of types.
+
+        2.  For each subpattern `e` in `p`:
+
+            1.  If `e` is a rest element pattern with a subpattern `s` and the
+                context type schema of `s` is an `Iterable<T>` for some type
+                `T`, then add `T` to `es`.
```

Comment:
Do we want to mention that `T` can be a type schema, since the schema of `s` can be `List<?>`.
In that case `T` is not a type, and isn't added to the list of types, and that's fine.
Might just want to be explicit about it.

Reply:
Ah, good catch. Done: https://github.com/dart-lang/language/commit/a93b4b14776eb0e802065c8eb39dd9c11d514fa5